### ISSUE TITLE
BgpUpdateMessageLayer: the value of `uint8_t prefix` of `struct prefix_and_ip` should not be limited to 8, 16, 24, 32

### ISSUE THREAD

**Issue by zanna-37** [seladb/PcapPlusPlus#631]
In the BGP protocol I noticed that in the `uint8_t prefix` of the `struct prefix_and_ip` the possible value is limited to 8, 16, 24, 32. ([docs ref.](https://github.com/seladb/PcapPlusPlus/blob/f94bbf8133a32033fcb80b765d543ce147f57a46/Packet%2B%2B/header/BgpLayer.h#L283-L284)) ([source code ref.](https://github.com/seladb/PcapPlusPlus/blob/f94bbf8133a32033fcb80b765d543ce147f57a46/Packet%2B%2B/src/BgpLayer.cpp#L379-L401))

> IPv4 address mask, must contain one of the values: 8, 16, 24, 32
    
On the contrary, in the RFC it seems to me that there is no such limitation. In fact, they even suggest a case in which the length is 0.

ref. [RFC page 15](https://tools.ietf.org/html/rfc4271#page-16):

>Length:
> The Length field indicates the length in bits of the IP
> address prefix.  A length of zero indicates a prefix that
> matches all IP addresses (with prefix, itself, of zero
> octets).

What do you think about it?

**Comment by seladb** [comment]
I believe you're right @zanna-37 , the values 8, 16, 24, 32 are the easiest to calculate in terms of IPv4 mask... Do you have a need to support all possible values or just add 0?

**Comment by zanna-37** [comment]
Well, I don't really need it. We are doing a university project where we are implementing BGP from scratch to run a simulator.
We are using your library (_and really thanks for your work_) for all the other packets but we are rewriting the BGP layer.
Once in a while, I check what you did in your implementation and I noticed that difference from the RFC (and also another minor regarding the maximum supported size for some optional parameters).

To sum up, no worries. It was just for letting you know. I will send you a ping here once our project is finished if you want to have a look at our implementation, but it would be probably too different to do a merge without a major refactor. And not even counting the fact that it might not be as high-quality as a library of this importance requires, but maybe worth having a look.

**Comment by seladb** [comment]
thanks for your kind words @zanna-37 ! I'm excited and humbled that you're using this library for your university project!
I'm curious why you're implementing the BGP layer from scratch, is it because of the project's requirement or is there anything missing in PcapPlusPlus BGP layer (besides the `prefix`)?
I'm very interested to see your work. If you implement a better version of the BGP layer, would you consider contributing it to this project? If so, maybe you can take the existing BGP layer and add/change what you need?

**Comment by zanna-37** [comment]
It is a soft requirement but we thought it was an interesting experience to get our hands dirty.
I'll ping you when it is finished so you can have a look and we can discuss how to proceed.
Thank you, have a nice day!

**Comment by seladb** [comment]
Thanks @zanna-37 ! Please feel free to ping when you're done.
I'll close this issue for now, please reopen if needed.
